cirq/protocols/resolve_parameters.py

```diff
@@ -55,6 +66,8 @@ def is_parameterized(val: Any) -> bool:
     """
     if isinstance(val, sympy.Basic):
         return True
+    if isinstance(val, numbers.Number):
```

Comment:
This is an interesting case which probably should be handled, but I'm not sure if this is the right place. What cases were you trying to capture? I don't see it covered in the tests.

I think this should come first. 

```
>>> isinstance(sp.sympify("1"), numbers.Number)
True
>>> isinstance(sp.sympify("1"), sp.Basic)
True
>>> isinstance(sp.sympify("a + 1"), sp.Basic)
True
>>> isinstance(sp.sympify("a + 1"), numbers.Number)
False
```

Reply:
Moved this to be the first thing checked before `sympy.Basic`. I don't recall exactly which failures this fixed, but some of the existing classes that implement the resolve parameters protocol did so incorrectly. This gets covered by some of the tests I've added that check for consistent implementation of the resolve parameters protocol, but I will add some more specific unit tests.

Reply:
That could be.  I think I am the one who added this to parameter resolution so that it avoids sympy.subs which is slow as molasses.  The goal was that not too much should fall through to sympy's formula resolution for performance reasons.  It's totally possible I goofed and didn't implement sympy's typing correctly or in the right order.

Reply:
Hm, so putting the `Number` check first results in some failures because we have code that assumes that if code is not parameterized then it's possible to do floating point operations on it. For example, in TwoQubitDiagonalGate if the gate is not parameterized we try to compute the unitary (https://github.com/quantumlib/Cirq/blob/master/cirq/ops/two_qubit_diagonal_gate.py#L63), but this fails if the gate contains sympy expressions even if there are no free parameters. I guess to model this properly we'd need separate checks for "is symbolic" versus "is parameterized". For now, I'm going to leave the number checks after the sympy checks.

Reply:
OK, added https://github.com/quantumlib/Cirq/issues/3310 for further work.